src/SymbolService.cpp

```diff
@@ -213,55 +218,70 @@ std::optional<SymbolInfo> SymbolService::locateSymbol(
       .debuginfo_path = &debuginfo_path,
   };
 
-  Dwfl* dwfl = dwfl_begin(&proc_callbacks);
+  dwfl = dwfl_begin(&proc_callbacks);
   if (dwfl == nullptr) {
     LOG(ERROR) << "dwfl_begin: " << dwfl_errmsg(dwfl_errno());
-    return std::nullopt;
+    return false;
   }
-  BOOST_SCOPE_EXIT_ALL(&) {
-    dwfl_end(dwfl);
-  };
 
-  switch (target.index()) {
-    case 0: {
-      auto pid = std::get<pid_t>(target);
-      if (int err = dwfl_linux_proc_report(dwfl, pid)) {
-        LOG(ERROR) << "dwfl_linux_proc_report: " << dwfl_errmsg(err);
-        return std::nullopt;
-      }
-      break;
-    }
-    case 1: {
-      const auto& exe = std::get<fs::path>(target);
-      Dwfl_Module* mod =
-          dwfl_report_offline(dwfl, exe.c_str(), exe.c_str(), -1);
-      if (mod == nullptr) {
-        LOG(ERROR) << "dwfl_report_offline: " << dwfl_errmsg(dwfl_errno());
-        return std::nullopt;
-      }
-
-      Dwarf_Addr start = 0;
-      Dwarf_Addr end = 0;
-      if (dwfl_module_info(mod, nullptr, &start, &end, nullptr, nullptr,
-                           nullptr, nullptr) == nullptr) {
-        LOG(ERROR) << "dwfl_module_info: " << dwfl_errmsg(dwfl_errno());
-        return std::nullopt;
-      }
-
-      VLOG(1) << "Module info for " << exe << ": start= " << std::hex << start
-              << ", end=" << end;
-
-      // Add module's boundary to executableAddrs
-      executableAddrs = {{start, end}};
-
-      break;
-    }
+  dwfl_report_begin(dwfl);
+
+  bool ok = std::visit(
+      visitor{[this](pid_t target) -> bool {
```

Comment:
Could we make these two methods with descriptive names? I initially had quite a hard time following this. E.g. `loadModulesFromPid` and `loadModulesFromExecutable`.

Reply:
I wish I didn't have to use `std::visit`, but it's the only type-safe way of ensuring all _variants_ are covered 😞